sdk/opa_test.go

```diff
@@ -588,6 +756,204 @@ main {
+	data.junk.x = input.y
+}
+`,
+		}),
+	)
+
+	defer server.Stop()
+
+	config := fmt.Sprintf(`{
+		"services": {
+			"test": {
+				"url": %q
+			}
+		},
+		"bundles": {
+			"test": {
+				"resource": "/bundles/bundle.tar.gz"
+			}
+		},
+		"decision_logs": {
+			"console": true
+		}
+	}`, server.URL())
+
+	opa, err := sdk.New(ctx, sdk.Options{
+		Config: strings.NewReader(config),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer opa.Stop(ctx)
+
+	m := metrics.New()
+	var result *sdk.PartialResult
+	if result, err = opa.Partial(ctx, sdk.PartialOptions{
+		Input:    map[string]int{"y": 2},
+		Query:    "data.test.allow = true",
+		Unknowns: []string{"data.junk.x"},
+		Mapper:   &sdk.RawMapper{},
+		Now:      time.Unix(0, 1619868194450288000).UTC(),
+		Metrics:  m,
+	}); err != nil {
+		t.Fatal(err)
+	} else if decision, ok := result.Result.(*rego.PartialQueries); !ok || decision.Queries[0].String() != "2 = data.junk.x" {
+		t.Fatal("expected &{[2 = data.junk.x] []} true but got:", decision, ok)
+	}
+
+	if exp, act := 5, len(m.All()); exp != act {
+		t.Fatalf("expected %d metrics, got %d", exp, act)
+	}
+
+	expectedRecordedMetrics := map[string]bool{
+		"timer_rego_input_parse_ns":   true,
+		"timer_rego_partial_eval_ns":  true,
+		"timer_rego_query_compile_ns": true,
+		"timer_rego_query_parse_ns":   true,
+		"timer_sdk_decision_eval_ns":  true,
+	}
+	for k := range m.All() {
+		if ok := expectedRecordedMetrics[k]; !ok {
+			t.Errorf("unexpected metric %s recorded", k)
+		}
+	}
+
+}
+
+func TestPartialWithInstrumentationAndProfile(t *testing.T) {
+
+	ctx := context.Background()
+
+	server := sdktest.MustNewServer(
+		sdktest.MockBundle("/bundles/bundle.tar.gz", map[string]string{
+			"main.rego": `
+package test
+
+allow {
+	data.junk.x = input.y
+}
+`,
+		}),
+	)
+
+	defer server.Stop()
+
+	config := fmt.Sprintf(`{
+		"services": {
+			"test": {
+				"url": %q
+			}
+		},
+		"bundles": {
+			"test": {
+				"resource": "/bundles/bundle.tar.gz"
+			}
+		},
+		"decision_logs": {
+			"console": true
+		}
+	}`, server.URL())
+
+	opa, err := sdk.New(ctx, sdk.Options{
+		Config: strings.NewReader(config),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer opa.Stop(ctx)
+
+	m := metrics.New()
+	p := profiler.New()
+	var result *sdk.PartialResult
+	if result, err = opa.Partial(ctx, sdk.PartialOptions{
+		Input:      map[string]int{"y": 2},
+		Query:      "data.test.allow = true",
+		Unknowns:   []string{"data.junk.x"},
+		Mapper:     &sdk.RawMapper{},
+		Now:        time.Unix(0, 1619868194450288000).UTC(),
+		Metrics:    m,
+		Profiler:   p,
+		Instrument: true,
+	}); err != nil {
+		t.Fatal(err)
+	} else if decision, ok := result.Result.(*rego.PartialQueries); !ok || decision.Queries[0].String() != "2 = data.junk.x" {
+		t.Fatal("expected &{[2 = data.junk.x] []} true but got:", decision, ok)
+	}
+
+	if exp, act := 32, len(m.All()); exp != act {
+		t.Fatalf("expected %d metrics, got %d", exp, act)
+	}
+
+	expectedRecordedMetrics := map[string]bool{
```

Comment:
This might be unnecessary. Perhaps we can get away with just checking the counts of the metrics or a subsample. Wdyt?

Reply:
I agree, reducing this is fine. An alternative approach would be to create a mock metrics implementation that merely counts calls to `Timer()` etc; but we'd get approximately the same assurance from that as we'd get from only checking counts here. 

I could imagine this test becoming annoying, because it's got nothing to do with compiler|topdown internals, but still asserts things so exactly that changing small things in topdown could cause this test to fail.

Perhaps we can check some lower bound? Some `x` for which we know that if `len(m.All())` is not greater-or-equal, then something is wrong. That would give us enough leeway to have changes in the compiler|topdown without having to update this test.

Reply:
>I could imagine this test becoming annoying

Indeed, keen to avoid that! 😄 

> Perhaps we can check some lower bound?

I think that might work, but if the default set of metrics were to grow over time above the lower bound number, might this test silently become meaningless?

I had another idea: 98605fd136fd5c48d37752b45f32c34a7bc734fb. Instead of looking for every single metric, we just check that certain collections of there are there, based on the presence of a prefixed metric. I've update all the tests on metrics to use that, we could even have a test helper if we wanted to DRY it up.